source/common/stats/thread_local_store.cc

```diff
@@ -137,6 +138,18 @@ void ThreadLocalStoreImpl::initializeThreading(Event::Dispatcher& main_thread_di
   tls_->set([](Event::Dispatcher&) -> ThreadLocal::ThreadLocalObjectSharedPtr {
     return std::make_shared<TlsCache>();
   });
+
+  // Initialize stats for each thread's dispatcher. This can't be done in the InitializeCb passed
+  // to `set` above, it has to be done as a separate step here, because initializing stats causes
+  // new TLS slots to be created and SlotImpl::set isn't reentrant.
+  auto worker_id = std::make_shared<std::atomic<uint32_t>>(0);
```

Comment:
is it possible for this to be deterministic on an ordered container in tls_ ? I could imagine interesting patterns of hot-spotting threads that this would obfuscate.

Reply:
I'm not sure if there's an easy way to do it based on TLS slots, but I could do this all a different way: require a name in `DispatcherImpl`'s ctor (which would have to flow through `Api::allocateDispatcher`). So the name would be set for the main thread in the server instance's ctor, and deterministically for each of the worker threads in `ProdWorkerFactory::createWorker`.